### PR TITLE
Story 15

### LINKED ISSUE #15: Encapsulate `EggsBatch` as a implementation detail
`EggsBatch` is an implementation detail and should not be exposed.
Create a package named `com.marianbastiurea.lifeofbees.eggframe` and place in it the `EggsBatch` and the `EggsFrame` classes. 
Add a `List<EggsBatch> eggBatches` field to the  `EggsFrame` class.
Make class `EggsBatch` package private and don't expose its details outside `EggsFrame` class.

Remove the `numberOfEggs` field  from  `EggsFrame`  class and instead expose a method that counts all the eggs in the `eggBatches` list and returns the total number of eggs in the frame.

### PR DESCRIPTION
**Pull request by MarianBastiurea**

### CODE REVIEW

File: src/main/java/com/marianbastiurea/lifeofbees/Apiary.java [hunk 1]
```diff
@@ -211,7 +201,6 @@ public void hibernate(Hive hive) {
         hive.setItWasSplit(false);
         hive.setAnswerIfWantToSplit(false);
         hive.setWasMovedAnEggsFrame(false);
-        hive.getEggsBatches().clear();
```
Comment (andreiursudev) [inline src/main/java/com/marianbastiurea/lifeofbees/Apiary.java:214]: Clear eggBatches from EggFrames instead.

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java [hunk 2]
```diff
@@ -515,4 +460,28 @@ public void moveAnEggsFrameFromUnsplitHiveToASplitOne() {
     public void changeQueen() {
         queen = new Queen(0);
     }
-}
\ No newline at end of file
+
+    public void fillUpEggsFrame(Date currentDate, int numberOfEggs) {
+
+        int maxEggPerFrame = 6400; // a frame have around 8500 cells. 75% more or less are used by the queen to lay eggs.
```
Comment (andreiursudev) [inline src/main/java/com/marianbastiurea/lifeofbees/Hive.java]: This is a property of the EggsFrame class, so move it to the EggsFrame class.

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java [hunk 3]
```diff
@@ -515,4 +460,28 @@ public void moveAnEggsFrameFromUnsplitHiveToASplitOne() {
...
+        // Remaining cells are fill up with honey or are damaged
+
+        int size = getEggsFrames().size() - this.getNumberOfFullEggsFrame();
+        //int remainingEggs = numberOfEggs % size;
+
+        if (size != 0) {
+            int numberOfEggsToPutInFrame = numberOfEggs / size;
+
+            for (EggsFrame eggsFrame : getEggsFrames()) {
+                if (eggsFrame.getNumberOfEggs() < maxEggPerFrame) {
```
Comment (andreiursudev) [inline src/main/java/com/marianbastiurea/lifeofbees/Hive.java]: Move this logic to EggsFrame class in the addEggs method.

File: src/main/java/com/marianbastiurea/lifeofbees/eggframe/EggsBatch.java [hunk 4]
```diff
@@ -2,10 +2,12 @@
 
 import java.util.Date;
 
+
 class EggsBatch {
```
Comment (andreiursudev) [inline src/main/java/com/marianbastiurea/lifeofbees/eggframe/EggsBatch.java]: Rename class to EggBatch

File: src/main/java/com/marianbastiurea/lifeofbees/eggframe/EggsFrame.java [hunk 5]
```diff
@@ -1,8 +1,12 @@
...
 
 
+import com.marianbastiurea.lifeofbees.BeesBatch;
+
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Date;
+import java.util.*;
 
 public class EggsFrame {
```
Comment (andreiursudev) [inline src/main/java/com/marianbastiurea/lifeofbees/eggframe/EggsFrame.java]: Rename class to EggFrame

**Review by andreiursudev — CHANGES_REQUESTED** [review]

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java (hunk 2 above)
Comment (MarianBastiurea) [reply src/main/java/com/marianbastiurea/lifeofbees/Hive.java]: done

Comment (MarianBastiurea) [reply src/main/java/com/marianbastiurea/lifeofbees/Hive.java]: delete all maxEggPerFrame from all Hive class

File: src/main/java/com/marianbastiurea/lifeofbees/eggframe/EggsBatch.java (hunk 4 above)
Comment (MarianBastiurea) [reply src/main/java/com/marianbastiurea/lifeofbees/eggframe/EggsBatch.java]: done

File: src/main/java/com/marianbastiurea/lifeofbees/eggframe/EggsFrame.java (hunk 5 above)
Comment (MarianBastiurea) [reply src/main/java/com/marianbastiurea/lifeofbees/eggframe/EggsFrame.java]: done

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java (hunk 3 above)
Comment (MarianBastiurea) [reply src/main/java/com/marianbastiurea/lifeofbees/Hive.java]: done

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java [hunk 6]
```diff
@@ -296,12 +297,12 @@ public void addNewEggsFrameInHive(int year) {
...
+        for (int i = 0; i < eggFrames.size(); i++) {
+            if (this.eggFrames.get(i).getNumberOfEggs() == this.eggFrames.get(i).getMaxEggPerFrame() && eggsFrameFull < eggFrames.size()) {
                 eggsFrameFull += 1;
             }
         }
-        if (eggsFrameFull == eggsFrames.size()) {
+        if (eggsFrameFull == eggFrames.size()) {
             int maximumNumberOfFramesToAdd = 6 - eggsFrameFull;
             if (maximumNumberOfFramesToAdd != 0) {
                 this.addEggsFrames(createNewEggsFrame());
```
Comment (andreiursudev) [inline src/main/java/com/marianbastiurea/lifeofbees/Hive.java]: Remove `createNewEggsFrame` and `addEggsFrames` methods and just use `this.eggFrames.add(new EggFrame());`

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java [hunk 7]
```diff
@@ -401,12 +402,11 @@ public void beesDie(Date currentDate) {
 
     public boolean checkIfAll6EggsFrameAre80PercentFull() {
```
Comment (andreiursudev) [inline src/main/java/com/marianbastiurea/lifeofbees/Hive.java:391]: This method looks too complicated or wrong. Fix or simplify.

**Review by andreiursudev — CHANGES_REQUESTED** [review]

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java (hunk 3 above)
Comment (andreiursudev) [reply src/main/java/com/marianbastiurea/lifeofbees/Hive.java]: This is not done

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java [hunk 8]
```diff
@@ -351,12 +297,12 @@ public void addNewEggsFrameInHive(int year) {
 
 
         int eggsFrameFull = 0;
-        for (int i = 0; i < eggsFrames.size(); i++) {
-            if (this.eggsFrames.get(i).getNumberOfEggs() == maxEggPerFrame && eggsFrameFull < eggsFrames.size()) {
+        for (int i = 0; i < eggFrames.size(); i++) {
+            if (this.eggFrames.get(i).getNumberOfEggs() == this.eggFrames.get(i).getMaxEggPerFrame() && eggsFrameFull < eggFrames.size()) {
```
Comment (andreiursudev) [inline src/main/java/com/marianbastiurea/lifeofbees/Hive.java]: Move logic `this.eggFrames.get(i).getNumberOfEggs() == this.eggFrames.get(i).getMaxEggPerFrame()` in new method named `isEggFrameFull(){ return getNumberOfEggs() == getMaxEggPerFrame()` in EggFrame class

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java [hunk 9]
```diff
@@ -474,9 +419,8 @@ public boolean checkIfAll6EggsFrameAre80PercentFull() {
 
     public int getNumberOfFullEggsFrame() {
         int eggsFrameFull = 0;
-        int maxEggPerFrame = 6400;
-        for (int i = 0; i < eggsFrames.size(); i++) {
-            if (this.eggsFrames.get(i).getNumberOfEggs() == maxEggPerFrame && eggsFrameFull < eggsFrames.size()) {
+        for (int i = 0; i < eggFrames.size(); i++) {
+            if (this.eggFrames.get(i).getNumberOfEggs() == this.eggFrames.get(i).getMaxEggPerFrame() && eggsFrameFull < eggFrames.size()) {
```
Comment (andreiursudev) [inline src/main/java/com/marianbastiurea/lifeofbees/Hive.java]: Use `isEggFrameFull` method.

**Review by andreiursudev — CHANGES_REQUESTED** [review]

File: src/main/java/com/marianbastiurea/lifeofbees/eggframe/EggFrame.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+    public EggFrame() {
+        this.eggBatches = new ArrayList<>();
+    }
+
+    public int getNumberOfEggs() {
+        return eggBatches.stream().mapToInt(EggBatch::getNumberOfEggs).sum();
+    }
+
+    public int getMaxEggPerFrame() {
```
Comment (andreiursudev) [inline src/main/java/com/marianbastiurea/lifeofbees/eggframe/EggFrame.java:25]: Remove this method.

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java (hunk 6 above)
Comment (MarianBastiurea) [reply src/main/java/com/marianbastiurea/lifeofbees/Hive.java]: done

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java (hunk 7 above)
Comment (MarianBastiurea) [reply src/main/java/com/marianbastiurea/lifeofbees/Hive.java:391]: done

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java [hunk 11]
```diff
@@ -400,20 +400,37 @@ public void beesDie(Date currentDate) {
...
-                }
-            }
+        if (eggFrames.size() != 6) {
+            return false;
         }
-        if (eggsFrameFull == 6) {
-            allEggsFrameAreFull = true;
+
+        for (EggFrame eggFrame : eggFrames) {
+            if (eggFrame.getNumberOfEggs() <= 0.8 * eggFrame.getMaxEggPerFrame()) {
```
Comment (andreiursudev) [inline src/main/java/com/marianbastiurea/lifeofbees/Hive.java]: Move `eggFrame.getNumberOfEggs() <= 0.8 * eggFrame.getMaxEggPerFrame()` logic to new `is80PercentFull` method in `EggFrame` class.

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java [hunk 12]
```diff
@@ -400,20 +400,37 @@ public void beesDie(Date currentDate) {
...
-        return allEggsFrameAreFull;
+
+        return true;
+
+
+
+
+
+
+//        int eggsFrameFull = 0;
```
Comment (andreiursudev) [inline src/main/java/com/marianbastiurea/lifeofbees/Hive.java]: Remove comment out code.

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java [hunk 13]
```diff
@@ -351,15 +296,16 @@ public void addNewEggsFrameInHive(int year) {
 
 
         int eggsFrameFull = 0;
-        for (int i = 0; i < eggsFrames.size(); i++) {
-            if (this.eggsFrames.get(i).getNumberOfEggs() == maxEggPerFrame && eggsFrameFull < eggsFrames.size()) {
+        for (int i = 0; i < eggFrames.size(); i++) {
```
Comment (andreiursudev) [inline src/main/java/com/marianbastiurea/lifeofbees/Hive.java]: Use enhanced for loop.

File: src/main/java/com/marianbastiurea/lifeofbees/Hive.java [hunk 14]
```diff
@@ -455,28 +391,23 @@ public void beesDie(Date currentDate) {
...
+
+        return true;
     }
 
     public int getNumberOfFullEggsFrame() {
         int eggsFrameFull = 0;
-        int maxEggPerFrame = 6400;
-        for (int i = 0; i < eggsFrames.size(); i++) {
-            if (this.eggsFrames.get(i).getNumberOfEggs() == maxEggPerFrame && eggsFrameFull < eggsFrames.size()) {
+        for (int i = 0; i < eggFrames.size(); i++) {
```
Comment (andreiursudev) [inline src/main/java/com/marianbastiurea/lifeofbees/Hive.java]: Use enhanced for loop.

**Review by andreiursudev — CHANGES_REQUESTED** [review]

**Review by andreiursudev — APPROVED** [review]

**Pull request merged**